### DISCUSSION TITLE
python_type of JSONB

### DISCUSSION THREAD

**Discussion by amacfie-tc** [sqlalchemy/sqlalchemy#10177] [Usage Questions]
The [`python_type`](https://docs.sqlalchemy.org/en/20/core/type_api.html#sqlalchemy.types.TypeEngine.python_type) for JSONB columns in Postgres is `dict`. Shouldn't it be `JSONType` where
```python
JSONType = Union[str, int, float, bool, None, Mapping[str, "JSONType"], list["JSONType"]]
```
?

ETA: should be `dict`, not `Mapping`

**Comment by CaselIT** [comment]
Hi,

this is inherited from the `JSON` type. I guess dict is not really the correct value, but I think it's the most used.
What's the use case here @amacfie-tc ?

@zzzeek not sure if it may make sense changing this (maybe for v2.1)?

**Reply by amacfie-tc** [reply]
Use case is sqlacodegen where `python_type` is used to generate type annotations for models. Right now I have to run an extra script to replace `dict` with `JSONType`. I only use JSONB but presumably JSON should change too.

**Reply by CaselIT** [reply]
this maybe could something that can be implemented on the sqlcodegen side too.

**Reply by amacfie-tc** [reply]
Could be, although see https://github.com/agronholm/sqlacodegen/issues/212#issuecomment-1125386115

**Reply by CaselIT** [reply]
let's see mike opinion on changing the python type of json

**Reply by zzzeek** [reply]
why `list` but then `Mapping` and not `dict`?    every json I've ever seen uses a dict

**Reply by andrew222651** [reply]
Oh, yeah, I think it should be `dict`

**Reply by CaselIT** [reply]
so the idea may be to do this change for 2.1?

**Reply by zzzeek** [reply]
maybe we can change it now? not sure. what would break?

**Reply by CaselIT** [reply]
well everyone that uses python_type would break. like `type_.python_type(value)` would now break

**Reply by zzzeek** [reply]
sure but what are they using *that* for

**Reply by CaselIT** [reply]
no clue, that's why I suggested 2.1. It's not like it's urgent

**Reply by NixBiks** [reply]
Shouldn't it be a typed `dict` at least? E.g. `dict[Any, Any]` to make mypy happy
